cxx4opencl/address_spaces.txt

```diff
@@ -20,52 +20,25 @@ logic.
 
 ==== Casts
 
-C-style casts follow `OpenCL C v2.0 rules (s6.5.5)`. All cast operators
-permit conversion to generic address space implicitly. However converting
-from generic to named address spaces can only be done using
-`addrspace_cast`. Note that conversions between `__constant` and any other
-other address space are disallowed.
-
-TODO: Example with `addrspace_cast`.
-
-[[addrspace-deduction]]
-==== Deduction
-
-Address spaces are not deduced for:
-
-  * non-pointer/non-reference template parameters or any dependent types
-    except for template specializations.
-  * non-pointer/non-reference class members except for static data members
-    that are deduced to `__global` address space.
-  * non-pointer/non-reference alias declarations.
-  * decltype expressions.
+C-style casts follow rules of `OpenCL C v2.0 s6.5.5`. Conversions to generic
+address space can be made by any {cpp} cast operator (as an implicit
+conversion); converting from generic to named address space can only be done
+using dedicated `addrspace_cast` operator. Operator `addrspace_cast` can only
+convert between address spaces and no other conversions are allowed to occur.
+Note that conversions between `__constant` and any other other address space
+are disallowed.
 
 [source,c]
 ----------
-template <typename T>
-void foo() {
-  T m; // address space of m will be known at template instantiation time.
-  T * ptr; // ptr points to generic address space object.
-  T & ref = ...; // ref references an object in generic address space.
-};
-
-template <int N>
-struct S {
-  int i; // i has no address space.
-  static int ii; // ii is in global address space.
-  int * ptr; // ptr points to generic address space int.
-  int & ref = ...; // ref references int in generic address space.
-};
-
-template <int N>
-void bar()
-{
-  S<N> s; // s is in __private address space.
-}
+int * genptr;
+__private float * ptrfloat = reinterpret_cast<__private float*>(genptr); //illegal.
+__private int * ptr = addrspace_cast<__private int*>(genptr);
```

Comment:
```suggestion
__private int * ptr = addrspace_cast<__private int*>(genptr); // legal
```

Would make this clearer to me, it's also what OpenCL C spec does for these examples.